h/security/policy/combined.py

```diff
@@ -21,32 +20,22 @@ class SecurityPolicy(IdentityBasedPolicy):
     This delegates to various different policies depending on the situation.
     """
 
-    def __init__(self, proxy_auth=False):
-        """
-        Initialise a security policy.
-
-        :param proxy_auth: Replace the default `CookiePolicy` for the UI with
-            the `RemoteUserPolicy`.
-        """
-        self._bearer_token_policy = BearerTokenPolicy()
-        self._http_basic_auth_policy = AuthClientPolicy()
+    def __init__(self):
         self._identity_cache = RequestLocalCache(self._load_identity)
 
-        self._ui_policy = RemoteUserPolicy() if proxy_auth else CookiePolicy()
-
     def remember(self, request, userid, **kw):
         """Get the correct headers to remember the given user."""
```

Comment:
That might lead to weird behaviors in some situations. For example a token-authenticated request to an API endpoint that supports either cookie or token-based authentication. `identity()` would return the result from `BearerTokenPolicy` but the above `remember()` implementation would return the result from `CookiePolicy`. That would be at least counter-intuitive, and possibly a cause of bugs. I think we want to make it always stick to one subpolicy per request (as per your other feedback).

It could be implemented something like this instead:

```python
def remember(self, ...):
    if isinstance(self._chosen_subpolicy, CookiePolicy):
        return self._chosen_subpolicy.remember(...)

    return []
```

...or:

```python
def remember(self, ...):
    try:
        return self._chosen_subpolicy.remember(...)
    except AttributeError:
        return []
```


...but I think either way is overall more complicated than the current approach because you now just have two different algorithms one for `identity()` and one for `remember()` and `forget()`; you've encoded knowledge in `SecurityPolicy` about what sub-policies there are and which of them do and don't support `remember()` and `forget()` rather than it just treating them as an opaque list; and you've removed the possibility of having multiple cookie-based or multiple `remember()` and `forget()`-supporting policies. So I think I prefer the design where each sub-policy implements all of Pyramid's `ISecurityPolicy` including `remember()` and `forget()`, if necessary with stub implementations that just return `[]`.